### PR TITLE
Refactor ES exceptions

### PR DESCRIPTION
**Pull request by everaldorodrigo**
# Context
The application should raise an exception when there is a timeout connection when connecting to the Elasticsearch.

However, there is an error in the exception handling function:
```
AttributeError: 'ConnectionTimeout' object has no attribute 'error'
```

There bug is in this line:
https://github.com/biothings/biothings.api/blob/8a430d25c8f31e7d43c74317887cee5feab32fd6/biothings/web/query/pipeline.py#L134

# Solution
Refactor the exception handling function.

### CODE REVIEW

File: biothings/web/query/pipeline.py [hunk 1]
```diff
@@ -100,52 +106,47 @@ async def _(*args, **kwargs):
...
-        except elasticsearch.RequestError as exc:  # 400s
+        except RequestError as exc:
             raise QueryPipelineException(400, *_simplify_ES_exception(exc))
+        except NotFoundError as exc:
+            raise QueryPipelineException(404, *_simplify_ES_exception(exc))
+        except ConflictError as exc:
+            raise QueryPipelineException(409, *_simplify_ES_exception(exc))
+        except (AuthenticationException, AuthorizationException) as exc:
+            raise QueryPipelineException(403, *_simplify_ES_exception(exc))
+        except Exception as exc:
```
Comment (ctrl-schaff) [inline biothings/web/query/pipeline.py:129]: We're doing a lot here in the general case. If possible I would try to add a little bit more context. For example in this chunk 
```
                if error_type == "search_phase_execution_exception":
                    if "rejected execution" in reason:
                        raise QueryPipelineException(503)
```
Why does `search_phase_execution_exception` coupled with `rejected_execution` translate to a 503 error

File: biothings/web/query/pipeline.py [hunk 2]
```diff
@@ -100,52 +106,47 @@ async def _(*args, **kwargs):
         ) as exc:
             logger.debug("QueryPipelineInterrupt: %s", exc.data)
             raise QueryPipelineInterrupt(exc.data)
```
Comment (ctrl-schaff) [inline biothings/web/query/pipeline.py:108]: Linters giving the suggestion we can attach the original context to the exception for all of these if we think it's relevant.
`raise-missing-from: Consider explicitly re-raising using 'raise QueryPipelineInterrupt(exc.data) from exc'`

File: biothings/web/query/pipeline.py [hunk 3]
```diff
@@ -100,52 +106,47 @@ async def _(*args, **kwargs):
         ) as exc:
             logger.debug("QueryPipelineInterrupt: %s", exc.data)
```
Comment (ctrl-schaff) [inline biothings/web/query/pipeline.py:107]: We do log the exception here but no where else. Potential choice to add `logger.exception` for exceptions where we think that would be useful

File: tests/web/test_es_exceptions.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
+import pytest
```
Comment (ctrl-schaff) [inline tests/web/test_es_exceptions.py:1]: Looks like it covers the gambit. Good tests

**Review by ctrl-schaff — APPROVED** [review]
Left a couple of comments for things to think about, but nothings that would block merging. Looks good

**Review by jal347 — APPROVED** [review]
I agree with Johnathan's comment that maybe adding more logger.exceptions in the future could be useful.

File: biothings/web/query/pipeline.py [hunk 5]
```diff
@@ -100,52 +106,47 @@ async def _(*args, **kwargs):
...
+        except ConflictError as exc:
+            raise QueryPipelineException(409, *_simplify_ES_exception(exc))
+        except (AuthenticationException, AuthorizationException) as exc:
+            raise QueryPipelineException(403, *_simplify_ES_exception(exc))
+        except Exception as exc:
+            status_code = getattr(exc, 'status_code', None)
+            if status_code in (429, "N/A"):
+                raise QueryPipelineException(503)
 
-        # it seems like the managed Elasticsearch service by AWS
```
[the remark this answers is not in the document]
Comment (everaldorodrigo) [reply biothings/web/query/pipeline.py:125]: AWS is not offering Elasticsearch anymore. AWS offers similar products called OpenSearch and CloudSearch. The `biothings.api` is using Elasticsearch from https://www.elastic.co/

So, I think is a good idea to keep the part of the comment not related to the Elasticsearch service manage by AWS.

File: biothings/web/query/pipeline.py (hunk 2 above)
Comment (ctrl-schaff) [reply biothings/web/query/pipeline.py:108]: Understood that makes sense. What do you think would entail a refactor for our exception handling if we were to make a ticket about it?

File: biothings/web/query/pipeline.py (hunk 3 above)
Comment (ctrl-schaff) [reply biothings/web/query/pipeline.py]: Based off your previous comment that's perfectly fine by me

File: biothings/web/query/pipeline.py (hunk 2 above)
Comment (everaldorodrigo) [reply biothings/web/query/pipeline.py:108]: I think it's a great idea creating a ticket for this!

**Pull request merged**